NuRadioReco/modules/phasedarray/triggerSimulator.py

```diff
@@ -18,265 +19,284 @@
-                elif sampling_rate != det.get_channel(station_id, channel_id)["trigger_adc_sampling_frequency"]:
-                    value_error = True
-            else:
-                if sampling_rate is None:
-                    sampling_rate = channel.get_sampling_rate()
-                elif sampling_rate != channel.get_sampling_rate():
-                    value_error = True
-            if value_error:
-                error_msg = 'Phased array channels do not have matching sampling rates. '
-                error_msg += 'Please specify a common sampling rate.'
-                raise ValueError(error_msg)
-
-        if trigger_adc and (upsampling_factor is not None):
-            upsampling_factor = int(upsampling_factor)
+def get_beam_rolls(station, det, triggered_channels,
+                   phasing_angles=default_angles, ref_index=1.55,
+                   trigger_adc=False, upsampling_factor=None):
+    """
+    Calculates the delays needed for phasing the array.
+    """
+    station_id = station.get_id()
+    sampling_rate = None
+    value_error = False
+
+    for channel_id in triggered_channels:
+        channel = station.get_channel(channel_id)
+        if trigger_adc:
+            if sampling_rate is None:
+                sampling_rate = det.get_channel(station_id, channel_id)["trigger_adc_sampling_frequency"]
+            elif sampling_rate != det.get_channel(station_id, channel_id)["trigger_adc_sampling_frequency"]:
+                value_error = True
         else:
-            upsampling_factor = 1
+            if sampling_rate is None:
+                sampling_rate = channel.get_sampling_rate()
+            elif sampling_rate != channel.get_sampling_rate():
+                value_error = True
+        if value_error:
+            error_msg = 'Phased array channels do not have matching sampling rates. '
+            error_msg += 'Please specify a common sampling rate.'
+            raise ValueError(error_msg)
+
+    if trigger_adc and (upsampling_factor is not None):
+        upsampling_factor = int(upsampling_factor)
+    else:
+        upsampling_factor = 1
+
+    time_step = 1. / sampling_rate / upsampling_factor
+
+    ant_z = get_antenna_positions(station, det, triggered_channels, 2)
+    check_vertical_string(station, det, triggered_channels)
+    beam_rolls = []
+    ref_z = (np.max(ant_z) + np.min(ant_z)) / 2
+
+    for angle in phasing_angles:
+        subbeam_rolls = {}
+        for z, channel_id in zip(ant_z, triggered_channels):
+            delay = (z - ref_z) / cspeed * ref_index * np.sin(angle)
+            roll = int(np.round(delay / time_step))
+            subbeam_rolls[channel_id] = roll
+        logger.debug("angle:", angle / units.deg)
+        logger.debug(subbeam_rolls)
+        beam_rolls.append(subbeam_rolls)
+
+    return beam_rolls
+
+
+def get_channel_trace_start_time(station, triggered_channels):
```

Comment:
Needs docs. Also, I'm not sure I see the point of the get_channel_trace_start_time? An explanation about why this careful checking is needed would be good.